### DISCUSSION TITLE
OIIOTOOL arguments and their intended return object

### DISCUSSION THREAD

**Discussion by jakubjezek001** [ynput/LabLib#11] [Q&A]
**Currently we are facing the decision on what class object should be returning oiiotool/ffmpeg/ocio arguments. Should it be returned only from Processors, or could it also be returned from Operators?**

One viewpoint is that Processors are responsible for handling the main tasks and operations, and therefore, should be the only ones to return a tool argumens. On the other hand, some argue that Operators also play an important role in the overall process and should be allowed to return the class object.

It is essential for us to carefully consider the implications and make a well-informed decision on this matter. This would not only affect the current project but also the future development and maintenance of the code.

What are your thoughts on this? Which approach do you think would be more suitable for our project? We would appreciate any input and suggestions from fellow engineers. Let's discuss and come to a consensus on this matter. Thank you.

**Comment by tweak-wtf** [comment]
while working on the codebase it somehow felt right to return command arguments from `operators` although i am pivoting more towards offloading this to `processors` as `operators` should really just be data objects.

however, if we implement `get_oiiotool_args` on `processors` i would suggest an inheritance structure for `operators` to support type hinting in `processors` since we would need to check for an `operator` classtype inorder to return the correct arguments.
this would mean more code to get that separation right. on the other hand side this separation is already made when return arguments from `operators`

**Comment by maxpareschi** [comment]
My two cents:
Operators are really just holders of data. The methods they can have should just really be functions that help either validate or process raw data passed to them to be stored in the correct format.

Processors are the main source of processing commands. It you have just one operator you can just pass that single operator to a processor and have your arguments formed that way.

This would help separate logic cleanly. If you are in need of custom processing it's probably better to build a BaseProcessor class which you can inherit from and overload/override methods on.

Also, processors should only be responsible for outputting the relevant part of the command line arguments, which then can be fed to the renderers.

Having custom logic on operators would lead to snowball the number of them and catch edge cases on processors probably...

Example, right now we are using internally just the effectprocessor with the relevant operators and a colorprocessor to generate ocio configs per shot and assign them to OCIO env with a prelaunch hook.

I did not touch operators directly, also because the color operators are a direct translation of OCIO classes.

All the custom logic is in the EffectProcessor, which collects the raw data from the effect file and saves that data on the color operators.

Then I pass that formed chain to the colorporcessor directly and save the ocioconfig from there, discarding the command line arguments as I have no use for them at the moment.

But I could have passed a manual list of color operators and it would have been the same result.

Point being, data on operators is always the same, so processors can do custom logic on top of a reliable data structure.
